### DISCUSSION TITLE
Migrate to User-Agent Client Hints

### DISCUSSION THREAD

**Discussion by Daniel15** [plausible/analytics#3420] [Anything Else]
Major browsers are freezing the User-Agent, meaning they'll have less fidelity and it won't be sufficient for identifying the browser. Over time, the user-agent will be deprecated.

The suggested replacement is to use Client Hints, which is an API that replaces the User-Agent. The server can request client hints by using a header, and there's also a JS API to request the client hints.

https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API

~~Plausible should migrate from using User-Agent to instead use client hints.~~
Plausible should consider using client hints in browsers that support them, falling back to User Agent in others.

**Comment by Nikola-Milovic** [comment]
> Experimental: This is an [experimental technology](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Experimental_deprecated_obsolete#experimental)
Check the [Browser compatibility table](https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API#browser_compatibility) carefully before using this in production.

This is far from _should_ migrate. It's something to keep an eye out for going forward, probably in a couple of years once the browser support is there. It's not even gonna leave experimental any time soon probably.

**Reply by Daniel15** [reply]
You're right. I should have said that Plausible should consider using client hints in browsers that support them, falling back to User Agent in others.

Google Chrome has already frozen the user agent and migrated to user agent hints, and accessing the user agent in JS logs a deprecation notice.

One of the benefits of user agent client hints is that you don't need complicated code to parse the user-agent, since each piece of information is a separate field.

**Comment by zemlanin** [comment]
FWIW, UA Client Hints [seem](https://github.com/faisalman/ua-parser-js/pull/696) to be the only reliable way to distinguish "Edge on Xbox" from "Edge on Windows" — the browser has the same user-agent string on different platforms

**Reply by Daniel15** [reply]
That's likely going to be the case for more hardware in the future. AFAIK neither Chrome nor Firefox are planning on adding new User-Agent strings.
